### PR TITLE
New Adapter: TradPlus

### PR DESCRIPTION
**Pull request by tradplus**
TradPlus is a leading, stable, and reliable ad monetization platform that is committed to providing global developers with fair, transparent and efficient monetization solutions. We also provide ADX and Saas ADX services. We serve 2,000+ global developers. Moreover, Our SDK has been certified by IAB Tech Lab and also joins Google Play SDK Index. And we have Information Security Management Certification.

Our advertising business spans the world. This is our official website: https://tradplusad.com/en. You can get more details about TradPlus.

this is TradPlus adapter version 1.0.

Thank!

Note: TradPlus only looks at the ext for the first impression and ignored the rest.

### CODE REVIEW

File: adapters/tradplus/params_test.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	`{"accountId": "11233", "zoneId": ""}`,
+	`{"accountId": "aaa", "accountId": "us"}`,
+	`{"accountId": "aa", "accountId": "sin"}`,
+}
+
+var invalidParams = []string{
+	`{"accountId": ""}`,
+	`{"accountId": "", "zoneId": ""}`,
+	`{"accountId": "", "zoneId": "sin"}`,
+	`{"zoneId": "aaa"}`,
```
Comment (SyntaxNode) [inline adapters/tradplus/params_test.go:47]: Please add test cases for an invalid data type, such as a number, boolean, or object for where a string is expected. Please also add test cases for a null string.

File: adapters/tradplus/tradplus.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtImpTradPlus) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{
+		AccountID: params.AccountID,
+		ZoneID:    params.ZoneID,
+	}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func transform(request *openrtb2.BidRequest) error {
```
Comment (SyntaxNode) [inline adapters/tradplus/tradplus.go:108]: Recommend to choose a more specific name. such as transformToNative11 or something similar to make the functionality clear.

File: adapters/tradplus/tradplustest/exemplary/simple-native-1.1.json [hunk 3]
```diff
@@ -0,0 +1,78 @@
+{
+  "mockBidRequest": {
+    "id": "test-request-id",
+    "imp": [
+      {
+        "id": "test-imp-id",
+        "native": {
+          "request": "{\"native\":{\"ver\":\"1.2\",\"context\":1,\"plcmttype\":4,\"plcmtcnt\":1,\"assets\":[{\"id\":2,\"required\":1,\"title\":{\"len\":90}},{\"id\":6,\"required\":1,\"img\":{\"type\":3,\"wmin\":128,\"hmin\":128,\"mimes\":[\"image/jpg\",\"image/jpeg\",\"image/png\"]}},{\"id\":7,\"required\":1,\"data\":{\"type\":2,\"len\":120}}]}}",
```
Comment (SyntaxNode) [inline adapters/tradplus/tradplustest/exemplary/simple-native-1.1.json:8]: Does it matter for your server that the ver is 1.2 when expected to be 1.1?

File: adapters/tradplus/tradplus.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	adapterRequest, errs := a.makeRequest(request)
+	if errs != nil {
+		return nil, errs
+	}
+	return []*adapters.RequestData{adapterRequest}, nil
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, []error) {
+
+	tradplusExt, err := getImpressionExt(&request.Imp[0])
```
Comment (SyntaxNode) [inline adapters/tradplus/tradplus.go:43]: Please link to your prebid.github.io PR and make it clear in that PR that the TradPlus only looks at the ext for the first impression and ignored the rest. Some other adapters do it, but its not a publisher expected behavior.

File: adapters/tradplus/params_test.go (hunk 1 above)
Comment (tradplus) [reply adapters/tradplus/params_test.go:51]: done.

File: adapters/tradplus/tradplus.go (hunk 2 above)
Comment (tradplus) [reply adapters/tradplus/tradplus.go]: This function has been removed, and supported in our service.

File: adapters/tradplus/tradplustest/exemplary/simple-native-1.1.json (hunk 3 above)
Comment (tradplus) [reply adapters/tradplus/tradplustest/exemplary/simple-native-1.1.json]: done.

File: adapters/tradplus/tradplus.go (hunk 4 above)
Comment (tradplus) [reply adapters/tradplus/tradplus.go:43]: ok，I've added to the first comment if anything else is needed ?

**Comment by tradplus** [comment]
@SyntaxNode , I have changed these codes，please review these codes.

File: adapters/tradplus/tradplus.go (hunk 4 above)
Comment (SyntaxNode) [reply adapters/tradplus/tradplus.go:43]: Yes. New adapters must also open a PR against the Prebid Docs repository here: https://github.com/prebid/prebid.github.io/

Please see the [User Documentation](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#user-documentation) section of our new adapter developer guide.

**Comment by tradplus** [comment]
@SyntaxNode I re-merged from master.

**Comment by bsardo** [comment]
Hi @tradplus, we recently released PBS 3.0, more specifically v3.1.0, which updates Prebid Server package import references throughout the project from `v2` to `v3`.
For example:
```
import (
    "github.com/prebid/prebid-server/v3/adapters"
)
```
As a result, please merge with master (no rebase) and then ensure all Prebid Server package import references in the files you’ve changed are `v3` such that the test suite passes so we can resume reviewing. Thanks!

File: adapters/tradplus/tradplus.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		return nil, &errortypes.BadInput{
+			Message: "Error parsing bidderExt - " + err.Error(),
+		}
+	}
+
+	if tradplusExt.AccountID == "" {
+		return nil, &errortypes.BadInput{
+			Message: "imp.ext.accountId required",
+		}
+	}
```
Comment (VeronikaSolovei9) [inline adapters/tradplus/tradplus.go:91]: this check is not needed here. It's handled by PBS core using properties set up:
```json
"accountId": {
      "type": "string",
      "description": "Account ID",
      "minLength": 1
    },
```

File: adapters/tradplus/tradplus.go (hunk 4 above)
Comment (VeronikaSolovei9) [reply adapters/tradplus/tradplus.go:44]: Follow up question: adapter code takes `imp[0].ext` and uses it to build endpoint url. Then sets` imp[0].ext` to nil, L49: `request.Imp[0].Ext = nil`
Other impressions are passed as is, with imp[n].ext. 
Is this expected? Do you want to set every imp[n].ext to nil?

File: adapters/tradplus/tradplus.go (hunk 5 above)
Comment (tradplus) [reply adapters/tradplus/tradplus.go]: done.

**Comment by VeronikaSolovei9** [comment]
Thank you for the changes. 
Please don't forget to add a documentation PR per [this comment](https://github.com/prebid/prebid-server/pull/3987#discussion_r1824407724). And decide if you want [to set all imps ext to nil ](https://github.com/prebid/prebid-server/pull/3987#discussion_r1889323463).

File: adapters/tradplus/tradplus.go (hunk 4 above)
Comment (tradplus) [reply adapters/tradplus/tradplus.go:44]: I've added a documentation PR: https://github.com/prebid/prebid.github.io/pull/5794.

Comment (tradplus) [reply adapters/tradplus/tradplus.go:44]: and we have adjusted the request.Imp[0].Ext logic so that it does not need to be set to nil.
thanks.

**Comment by tradplus** [comment]
@VeronikaSolovei9 done, thanks for the reminder. https://github.com/prebid/prebid-server/pull/3987#issuecomment-2578810383

**Comment by VeronikaSolovei9** [comment]
Thank you for the updates. Code looks good to me. Please pull latest changes from the main branch into your branch, this should fix the validate jobs.

**Review by VeronikaSolovei9 — DISMISSED** [review]
LGTM

File: adapters/tradplus/params_test.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		if err := validator.Validate(openrtb_ext.BidderTradPlus, json.RawMessage(invalidParam)); err == nil {
+			t.Errorf("Schema allowed unexpected params: %s", invalidParam)
+		}
+	}
+}
+
+var validParams = []string{
+	`{"accountId": "11233", "zoneId": ""}`,
+	`{"accountId": "aaa", "accountId": "us"}`,
+	`{"accountId": "aa", "accountId": "sin"}`,
```
Comment (guscarreon) [inline adapters/tradplus/params_test.go:40]: Lines 39 and 40 seem to test the exact same scenario. Did you mean to have a `zoneId` as parameter in one of them?
```
37   var validParams = []string{
38       `{"accountId": "11233", "zoneId": ""}`,
39 -     `{"accountId": "aaa", "accountId": "us"}`,
   +     `{"accountId": "aaa", "zoneId": "us"}`,
40       `{"accountId": "aa", "accountId": "sin"}`,
41   }
adapters/tradplus/params_test.go
```

**Review by guscarreon — COMMENTED** [review]
Can we add a JSON test case were an incoming request comes with more than one element in the `request.Imp[]` array? I believe it'll help to assert that elements from `request.Imp[1]` onward get ignored.

File: adapters/tradplus/params_test.go (hunk 6 above)
Comment (tradplus) [reply adapters/tradplus/params_test.go:40]: @guscarreon  yes, thanks for the reminder, it has been modified. [feat: add multiple imp tests](https://github.com/prebid/prebid-server/pull/3987/commits/5d9625992f7ea96687df469612fa547e0fa8562c)

**Review by guscarreon — APPROVED** [review]
LGTM

**Review by VeronikaSolovei9 — APPROVED** [review]
